### ISSUE TITLE
Which server types should be prioritised for Level 5?

### ISSUE THREAD

**Issue by hypernuggets** [nus-cs2030/2021-s2#230]
Hi, I'm trying to rework my method to get the shortest server for greedy customers at level 5. I'm a bit confused about the priority in assigning customers to each server.

Here are the possible server states that I think we're working with:

1 Server is not resting and has no customers
2 Serving is not resting and has customers
3 Serving is resting and has no customers
4 Serving is resting and has customers

If I'm not wrong, the order should be 1 3 2 4. Is this correct?

**Comment by OuaisBien** [comment]
Accurate. But I don't think you have to modify your method tho if you manage to pass all the test cases until level5.

**Comment by e0564330** [comment]
Hello! Not too sure if this is the correct approach but for the priority of assigning customers, I just followed the same priority provided from level 3 regardless of whether a server is resting or not resting(which sounds like the order you mentioned above), I mainly just made sure that a server that is resting can still take in customers into their queue. Hope this helps!

**Comment by eryuanren** [comment]
Hi! I think it is actually alot simpler than that. The first priority is always if the server is able to serve (both has no customers and not resting). If there is none, then check who has the shortest queue (regardless of whether he is resting)

Do take note the maximum queue length is the maximum number of customers allowed to wait at a server. That is to say, a server who is serving and has 2 customers waiting has the same queue length as a server who is resting and has 2 customers waiting

**Comment by markwty** [comment]
The logic for the first 2 in the order is correct, but the logic for 2 and 4 depends on case by case basis. So it can be 1 3 2 4 or 1 3 4 2 with the last 2 depending on the number of people in the queue i.e. a server that is resting may have a longer queue than a server that is resting.

**Comment by Xeph-o** [comment]
2 or 3 could swap depending on whether the customer is greedy or not. A non-greedy customer would go to 2 instead of 3 (assuming not full), while a greedy one will always go to the 3rd one instead of 2nd (shortest queue).

**Comment by hypernuggets** [comment]
> Hi! I think it is actually alot simpler than that. The first priority is always if the server is able to serve (both has no customers and not resting). If there is none, then check who has available space in the queue (regardless of whether he is resting)
> 
> Do take note the maximum queue length is the maximum number of customers allowed to wait at a server. That is to say, a server who is serving and has 2 customers waiting has the same queue length as a server who is resting and has 2 customers waiting

For anyone facing the same problem, I managed to get past it and it's pretty much this!

The flow of the checks that I used was as follows:

- Check for any empty Servers (i.e. no customers at all) 
- If there are empty servers, return the **first non-resting empty server**
- If none of the servers are empty, set the first available server **(regardless of resting or not)** as the Server result to check other servers' queue length against
- If any servers have shorter queue lengths, change the comparing server to that new server (using <, not <= to check queue lengths. this is so that if two servers have the same queue lengths, you're taking the first server in order of the servers)
- Return the server at the end

**Comment by gableejh** [comment]
when you say no customer, i assume you mean that the queue is empty? if so, then 1 and 3 is correct, but 2 or 4 depends on the length of the queue. for greedy customers, they will go for the shortest queue. but for normal customers, they'll just take a first queue available

**Comment by hypernuggets** [comment]
> when you say no customer, i assume you mean that the queue is empty? if so, then 1 and 3 is correct, but 2 or 4 depends on the length of the queue. for greedy customers, they will go for the shortest queue. but for normal customers, they'll just take a first queue available

Correct, empty servers means that the server has zero customers in the queue (regardless of serving or resting). When it comes to 2 or 4, whichever server has fewer customers/comes first if same number of customers will be selected

**Comment by KaveenE** [comment]
Can I confirm that when you say this:

> * If there are empty servers, return the **first non-resting empty server**
> * If none of the servers are empty, set the first available server **(regardless of resting or not)** as the Server result to check other servers' queue length against
> * If any servers have shorter queue lengths, change the comparing server to that new server (using <, not <= to check queue lengths. this is so that if two servers have the same queue lengths, you're taking the first server in order of the servers)
> * Return the server at the end

Since we only checked for a non-resting empty server initially, does this mean that when we check for the smallest queue(for greedy) regardless if the server is resting or not, it's possible for the queue to be empty?

**Comment by hypernuggets** [comment]
> Since we only checked for a non-resting empty server initially, does this mean that when we check for the smallest queue(for greedy) regardless if the server is resting or not, it's possible for the queue to be empty?

From my own implementation, it would be possible for the smallest queue server to be empty, in which case you would want to use that server since a queue length of 0 (when empty) is the shortest possible queue you can have. 

The reason for the first check is because if you have multiple empty servers (eg. servers 1 and 2 both have zero customers), you want to prioritise a non-resting server over a resting server.
